ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -99,6 +204,249 @@ public Integer getValue() {
+          break;
+        case InvalidBlobId:
+          invalidBlobIdErrorCount.inc();
+          break;
+        case InvalidPutArgument:
+          invalidPutArgumentErrorCount.inc();
+          break;
+        case OperationTimedOut:
+          operationTimedOutErrorCount.inc();
+          break;
+        case RouterClosed:
+          routerClosedErrorCount.inc();
+          break;
+        case UnexpectedInternalError:
+          unexpectedInternalErrorCount.inc();
+          break;
+        case BlobTooLarge:
+          blobTooLargeErrorCount.inc();
+          break;
+        case BadInputChannel:
+          blobTooLargeErrorCount.inc();
+          break;
+        case InsufficientCapacity:
+          insufficientCapacityErrorCount.inc();
+          break;
+        case BlobDeleted:
+          blobDeletedErrorCount.inc();
+          break;
+        case BlobDoesNotExist:
+          blobDoesNotExistErrorCount.inc();
+          break;
+        case BlobExpired:
+          blobExpiredErrorCount.inc();
+          break;
+        default:
+          unknownErrorCountForOperation.inc();
+          break;
+      }
+    } else {
+      unknownErrorCountForOperation.inc();
+    }
+  }
+
+  /**
+   * Get {@link NodeLevelMetrics} for a given {@link DataNodeId}. If the {@link DataNodeId} as the key does not exist,
+   * a {@link NodeLevelMetrics} corresponding to an unknown {@link DataNodeId} will be returned.
+   * @param dataNodeId The {@link DataNodeId} to be indexed.
+   * @return The {@link NodeLevelMetrics}.
+   */
+  NodeLevelMetrics getDataNodeBasedMetrics(DataNodeId dataNodeId) {
+    return dataNodeToMetrics.containsKey(dataNodeId) ? dataNodeToMetrics.get(dataNodeId) : dataNodeToMetrics.get(null);
+  }
+
+  /**
+   * A metrics class that tracks at the {@link DataNodeId} level. These metrics are collected based on the operation
+   * requests sent to individual {@link DataNodeId}. An operation request is part of an operation, and conveys an actual
+   * request to a {@link com.github.ambry.clustermap.ReplicaId} in a {@link DataNodeId}. An operation request can be
+   * either for a metadata blob, or for a datachunk.
+   */
+  public class NodeLevelMetrics {
+
+    // Request rate.
+    public final Meter putRequestRate;
+    public final Meter getBlobInfoRequestRate;
+    public final Meter getRequestRate;
+    public final Meter deleteRequestRate;
+
+    // Request latency.
+    public final Histogram putRequestLatencyMs;
+    public final Histogram getBlobInfoRequestLatencyMs;
+    public final Histogram getRequestLatencyMs;
+    public final Histogram deleteRequestLatencyMs;
+
+    // Request error count.
+    public final Counter putRequestErrorCount;
+    public final Counter getBlobInfoRequestErrorCount;
+    public final Counter getRequestErrorCount;
+    public final Counter deleteRequestErrorCount;
+
+    // Timed-out request count for each operation type.
+    public final Counter putRequestTimeoutCount;
```

Comment:
is this particularly useful? doesn't this get counted as part of the "request error count" above?